Dependent dashboard selectors

sashasashasasha151 opened the pull request:
Problem: it is possible to choose filters on dashboard page in a such way
that there will be no available experiments

Solution: filter available selector values based on other selectors

## Description

<!--
Describes the nature of your changes. If they are substantial, you should
further subdivide this into a section describing the problem you are solving and
another describing your solution.
-->

## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

https://issues.serokell.io/issue/EDNA-96

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

sashasashasasha151 committed:
Dependent dashboard selectors

Problem: it is possible to choose filters on dashboard page in a such way
that there will be no available experiments

Solution: filter available selector values based on other selectors

sashasashasasha151 committed:
All seletor items with delimeter

Problem: it is not possible to see all items it two selectors are chosen

Solution: add all selector items with delimeter and reset them if smth
incompatable is chosen

gromakovsky commented on frontend/src/pages/dashboard/EntitySelector.tsx:
```diff
@@ -123,24 +182,55 @@ export function CompoundSelector({ className }: SelectorProps): React.ReactEleme
+      return undefined;
+    },
+    [filterNeeds, experiments]
+  );
```
I am not a TS expert, but it looks like substantial copy-pasted. Did you try to define some helper to decrease code duplication?

gromakovsky replied:
Hm, ok, it's changed in the next commit and apparently doesn't look like too massive copy-paste now.

pva701 replied:
still looks like decent amount of copy-paste to me

gromakovsky commented:
Will you add the new Reset button here or in another issue?

sashasashasasha151 committed:
Add reset button

Problem: we need to be able to reset all selectors

Solution: create reset button

pva701 commented on frontend/src/components/Combobox.tsx:
```diff
@@ -20,6 +20,7 @@ interface ComboboxProps<T> extends FormikCompatible<Maybe<T>> {
   isLoading?: boolean;
   styles?: StylesConfig<SelectOption, false, GroupTypeBase<SelectOption>>;
   isDisabled?: boolean;
+  optionsFilter?: (val: T[]) => T[] | undefined;
```
are you sure you need both `?:` and `| undefined`? I don't think so

sashasashasasha151 replied:
It is not `((val: T[]) => T[]) | undefined`

pva701 commented on frontend/src/pages/dashboard/EntitySelector.tsx:
```diff
@@ -20,26 +21,123 @@ import {
+    experimentsF
+      .filter(
+        experiment =>
+          (isDefined(compound) ? experiment.compoundName === compound.item.name : true) &&
```
`!isDefined(compound) || experiment.compoundName === compound.item.name`

pva701 commented on frontend/src/pages/dashboard/EntitySelector.tsx:
```diff
@@ -20,26 +21,123 @@ import {
+          filteredProjects.push({
+            id: -1,
+            item: {
+              name: "Start new filter with:",
```
kak ya oru, it's just brilliant how you workarounded it :pepe_laugh_dead:

pva701 commented on frontend/src/pages/dashboard/EntitySelector.tsx:
```diff
@@ -20,26 +21,123 @@ import {
+
+        const [filteredProjects, restProjects] = projects.reduce(
+          (acc: ProjectDto[][], project) => {
+            acc[projectNames.has(project.item.name) ? 0 : 1].push(project);
```
honestly, looks like hujita. you could pattern-match on acc like:
`[filteredProjects, restProjects]` and make an if instead of this creepy statement
